cgmes/cgmes-conversion/src/main/java/com/powsybl/cgmes/conversion/SubstationIdMapping.java

```diff
@@ -10,70 +10,302 @@
+        if (voltageLevelMapping.containsKey(vlid)) {
+            return voltageLevelMapping.get(vlid);
+        }
+        return vlid;
+    }
+
+    // All the keys for a given value, all the merged voltageLevels that have cgmesIdentifier as representative
+    public List<String> mergedVoltageLevels(String cgmesIdentifier) {
+        String vlid = context.namingStrategy().getId(CgmesNames.VOLTAGE_LEVEL, cgmesIdentifier);
+        return voltageLevelMapping.entrySet().stream().filter(record -> record.getValue().equals(vlid))
+            .map(Map.Entry::getKey).collect(Collectors.toList());
+    }
+
     public void build() {
-        // CGMES standard:
-        // "a PowerTransformer is contained in one Substation but it can connect a Terminal to
-        // another Substation"
-        // Ends of transformers need to be in the same substation in the IIDM model.
-        // We will map some CGMES substations to a single IIDM substation
-        // when they are connected by transformers,
-        // that is, when there are at least one power transformer that has terminals on both
-        // substations
-
-        Graph<String, Object> g = graphSubstationsTransformers();
-        new ConnectivityInspector<>(g).connectedSets().stream()
-                .filter(substationIds -> substationIds.size() > 1)
-                .forEach(substationIds -> {
-                    String selectedSubstationId = representativeSubstationId(substationIds);
-                    for (String substationId : substationIds) {
-                        if (!substationId.equals(selectedSubstationId)) {
-                            mapping.put(substationId, selectedSubstationId);
-                        }
+        Map<String, List<String>> voltageLevelAdjacency = new HashMap<>();
+        Map<String, List<String>> substationAdjacency = new HashMap<>();
+
+        buildAdjacency(voltageLevelAdjacency, substationAdjacency);
+        buildVoltageLevel(voltageLevelAdjacency);
+        buildSubstation(substationAdjacency);
+    }
+
+    private void buildAdjacency(Map<String, List<String>> voltageLevelAdjacency,
+        Map<String, List<String>> substationAdjacency) {
+        context.cgmes().voltageLevels().forEach(vl -> addVoltageLevel(voltageLevelAdjacency, vl));
+        context.cgmes().substations().forEach(st -> addSubstation(substationAdjacency, st));
+
+        context.cgmes().switches().forEach(sw -> addSwitch(voltageLevelAdjacency, substationAdjacency, sw));
+        context.cgmes().groupedTransformerEnds().forEach((t, tends) -> addEnds(substationAdjacency, tends));
+    }
+
+    private void addVoltageLevel(Map<String, List<String>> voltageLevelAdjacency, PropertyBag vl) {
+        String voltageLevelId = vl.getId(CgmesNames.VOLTAGE_LEVEL);
+        String vId = context.namingStrategy().getId(CgmesNames.VOLTAGE_LEVEL, voltageLevelId);
+        voltageLevelAdjacency.computeIfAbsent(vId, k -> new ArrayList<>());
```

Comment:
I would write `voltageLevelAdjacency.put(vid, new ArrayList<>());`
Here, we are sure that the key is absent, so it's not necessary to check it

Reply:
Done